Implement resizeable sidebar in grid

tatankat opened the pull request:
I you suggested in https://github.com/rsjtdrjgfuzkfg/thunderbird-experiments/issues/10#issuecomment-1555917865, I created a PR to resize the sidebar. What do you think?
It still needs a proper way to edit the gridTemplate. I can't find a better way than some regex magic. Although, you may wish to keep it simple. As it just works this way as long as there is no other addon (or newer Thunderbird version) playing with the gridTemplate.

tatankat committed:
Implement resizable sidebar in grid

rsjtdrjgfuzkfg commented on experiments/customui/parent.js:
```diff
@@ -349,17 +350,35 @@ var ex_customui = class extends ExtensionCommon.ExtensionAPI {
+          let parentGridTemplate = window.getComputedStyle(container).gridTemplate;
+          container.style.setProperty(`--${splitter.id}-width`,width+"px");
+          //TODO generate this edit: add 2 columns
+          container.style.gridTemplate = `"folders folderPaneSplitter threads ${splitter.style.gridArea} ${sidebar.style.gridArea}" minmax(auto, 1fr) "folders folderPaneSplitter messagePaneSplitter ${splitter.style.gridArea} ${sidebar.style.gridArea}" min-content "folders folderPaneSplitter message ${splitter.style.gridArea} ${sidebar.style.gridArea}" minmax(auto, var(--messagePaneSplitter-height ...
```
This sounds very unstable. Imho the minimal change to make it reasonable for inclusion would be to do this change dynamically (as stated in the TODO).

rsjtdrjgfuzkfg commented on experiments/customui/parent.js:
```diff
@@ -368,9 +387,9 @@ var ex_customui = class extends ExtensionCommon.ExtensionAPI {
       // (if the largest requested width is too large); an implementation for
       // Core inclusion might want to address these issues.
       setWebextFrameDynamicVisibility(result, options);
-      setWebextFrameDynamicDimension(result, options, "width", 244);
```
Was setting the width through local options already broken? If not, this will most likely break that.

tatankat replied:
It was, you fixed it at 244. But the width given in options to insertSidebarWebextFrame still works. It is set as a css variable https://github.com/rsjtdrjgfuzkfg/thunderbird-experiments/blob/623747407ce1f5e098f242862031ab36680741bd/experiments/customui/parent.js#L377

rsjtdrjgfuzkfg commented on experiments/customui/parent.js:
```diff
@@ -383,6 +402,9 @@ var ex_customui = class extends ExtensionCommon.ExtensionAPI {
       const sidebar = removeWebextFrame(tag, url, document);
       if (sidebar && sidebar.childElementCount == 0) {
+        //TODO properly remove our edit, if it has been changed
+        // in case another addon also edited the gridTemplate
```
Similar to the one before, this TODO should be resolved before merging.

rsjtdrjgfuzkfg reviewed (changes requested):
Thanks for your PR!

It's been a while since I looked into this, so I'm unsure whether there would be an easier option to add the splitter. That said, adding things to grids you don't control is likely always a bit fragile. If you're willing to address the TODOs I think this could get to a state that can get merged.

In addition to your TODOs, you remove local resizing options: if that was not previously broken, it would be nice to preserve that feature.

----

Note that while this review states that I "request" changes, I'm unsure if those are worth your time, as CustomUI is effectively dead. Thunderbird's core maintainers decided against its concept of a generic UI extension interface; their plan is to add individual core APIs for each location (like a sidebar API for sidebars only).

If you have a bit of time on your hand and a desire for sidebars, you might thus want to skip working on CustomUI and instead use that time to contribute towards an official sidebar API that can at some point end up in core Thunderbird.

Imho it is no longer reasonable to invest a lot of time into CustomUI, but making adjustments to keep things running for add-ons that do not yet have official APIs can of course still make sense.

tatankat committed:
Dynamically changing the grid, fix layout changing

tatankat commented:
> Thanks for your PR!

My pleasure :) Thanks for looking into it.

> It's been a while since I looked into this, so I'm unsure whether there would be an easier option to add the splitter. That said, adding things to grids you don't control is likely always a bit fragile. If you're willing to address the TODOs I think this could get to a state that can get merged.

I addressed the TODOs. They were indeed needed as without it, it broke alternative layouts, see https://github.com/ganast/tidybird/issues/102.

> In addition to your TODOs, you remove local resizing options: if that was not previously broken, it would be nice to preserve that feature.

Giving the width in options still works, see comment

> Note that while this review states that I "request" changes, I'm unsure if those are worth your time, as CustomUI is effectively dead. Thunderbird's core maintainers decided against its concept of a generic UI extension interface; their plan is to add individual core APIs for each location (like a sidebar API for sidebars only).
> 
> If you have a bit of time on your hand and a desire for sidebars, you might thus want to skip working on CustomUI and instead use that time to contribute towards an official sidebar API that can at some point end up in core Thunderbird.

If you can point me to some development going on, I would be happy to help and test

> Imho it is no longer reasonable to invest a lot of time into CustomUI, but making adjustments to keep things running for add-ons that do not yet have official APIs can of course still make sense.

Yeah, I let myself go, but mainly to fix issues to keep the current features working. So no problem if you don't feel like merging, I understand any objections - this becomes a rather big edit to fix such simple feature. This can simply be kept as known workaround.

PS: I also added a new feature to customui, just to keep things properly running: getting css variable values to follow thunderbird's interface colors. But I won't bother you with it, unless you want me to.

rsjtdrjgfuzkfg commented:
>> contribute towards an official sidebar API
>
> If you can point me to some development going on, I would be happy to help and test

There is no existing development I know of, I suggested you start something :) I'm sure @jobisoft can assist with coordination of all the relevant steps (he is the module owner of the Addon Support module in Thunderbird).


> This can simply be kept as known workaround.

I'll take you on that offer :)
I'll do a very brief review, but let's keep the PR open for now.

Everyone else: if you tested this PR and you feel it should get merged, feel free to chime in in order to get me to merge it at some point ;)

rsjtdrjgfuzkfg commented on experiments/customui/parent.js:
```diff
@@ -373,12 +575,31 @@ var ex_customui = class extends ExtensionCommon.ExtensionAPI {
+          };
+          updateGridTemplate();
+          const observer = new window.MutationObserver(updateGridTemplate);
+          window.customuiObserver = observer;
```
This looks dangerous in situations in which you have multiple sidebars (possibly in different add-ons using this experiment), as you could overwrite the observer set for the previous sidebar.

tatankat replied:
In fact, it's ok they share the same observer (for now) as it essentially does the same and in some cases, we only want it to run only once. Multiple sidebars are currently vertically stacked without issue, which may be desired behavior.
If there are more add-ons using this, it could probably be looked after to have optional vertical or horizontal stacking.

rsjtdrjgfuzkfg replied:
At the very least you'd leak the observer (as you don't disconnect the previous one). Either way, just prefix the variable with `context.contextId` and be done with it?

rsjtdrjgfuzkfg commented on experiments/customui/parent.js:
```diff
@@ -319,7 +319,209 @@ var ex_customui = class extends ExtensionCommon.ExtensionAPI {
+    }
+
+    const addColumnsToElementGridTemplate = function(container, window, columns) {
+      console.log("Adding columns to grid");
```
I'd prefer to avoid unconditional debug output. (also applies below)

rsjtdrjgfuzkfg commented on experiments/customui/parent.js:
```diff
@@ -319,7 +319,209 @@ var ex_customui = class extends ExtensionCommon.ExtensionAPI {
-    
+
+    // get the grid-template css rule of a container
+    const getGridTemplate = function(container,document) {
```
This does not take the element's style into account, correct?

If so, multiple add-ons doing this would cause issues? Maybe even multiple sidebars (I did not do any testing)?

tatankat replied:
I don't think I fully understand your remark. If I understand your question, then the answer is it does take the element's style into account. That is essentially what it does.
Also I don't see why this would cause an issue. Maybe the timing of multiple added sidebars, but as they are vertically stacked at the moment, this is not an issue either.

rsjtdrjgfuzkfg replied:
It checks style sheets, but not a `style` attribute on the element itself.

rsjtdrjgfuzkfg reviewed (changes requested):
As discussed, this is just a short review to give some feedback. Addressing this feedback would be necessary for a merge, but as we don't plan on merging at the moment: *please feel free to completely ignore my comments!*

I'm glad you the PR to this point, so we do have a good workaround for any add-on authors that need sidebar support in 115. Thanks a lot @tatankat!

tatankat committed:
Compatibility Update for Thunderbird 128

Co-authored-by: John Bieling <john.bieling@gmx.de>

tatankat committed:
Remove console log
